File: src/dto/infrastructure.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use super::common::{GitDetails, EntityValidityDetails};

/// Infrastructure definition representation
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Infrastructure {
    #[serde(rename = "accountId")]
    pub account_id: String,
    #[serde(rename = "orgIdentifier")]
    pub org_identifier: String,
    #[serde(rename = "projectIdentifier")]
    pub project_identifier: String,
    #[serde(rename = "environmentRef")]
    pub environment_ref: String,
    pub identifier: String,
    pub name: String,
    pub description: Option<String>,
    pub tags: Option<HashMap<String, String>>,
    #[serde(rename = "type")]
    pub infra_type: String,
    #[serde(rename = "deploymentType")]
    pub deployment_type: String,
    pub yaml: Option<String>,
    #[serde(rename = "gitDetails")]
    pub git_details: Option<GitDetails>,
    #[serde(rename = "entityValidityDetails")]
    pub entity_validity_details: Option<EntityValidityDetails>,
    #[serde(rename = "storeType")]
    pub store_type: Option<String>,
    #[serde(rename = "connectorRef")]
    pub connector_ref: Option<String>,
}

/// Infrastructure input for creation/update
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InfrastructureInput {
    pub identifier: String,
    pub name: String,
    pub description: Option<String>,
    #[serde(rename = "environmentRef")]
    pub environment_ref: String,
    pub tags: Option<HashMap<String, String>>,
    #[serde(rename = "type")]
    pub infra_type: String,
    #[serde(rename = "deploymentType")]
    pub deployment_type: String,
    pub yaml: Option<String>,
}

/// Infrastructure filter options
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InfrastructureFilterOptions {
    #[serde(rename = "infraIdentifiers")]
    pub infra_identifiers: Option<Vec<String>>,
    #[serde(rename = "infraNames")]
    pub infra_names: Option<Vec<String>>,
    #[serde(rename = "envIdentifiers")]
    pub env_identifiers: Option<Vec<String>>,
    pub description: Option<String>,
    #[serde(rename = "infraTypes")]
    pub infra_types: Option<Vec<String>>,
    #[serde(rename = "deploymentTypes")]
    pub deployment_types: Option<Vec<String>>,
    pub tags: Option<HashMap<String, String>>,
    #[serde(rename = "filterType")]
    pub filter_type: Option<String>,
}

/// Infrastructure YAML move request
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InfrastructureYamlMoveRequest {
    #[serde(rename = "moveConfigOperationType")]
    pub move_config_operation_type: String,
    #[serde(rename = "connectorRef")]
    pub connector_ref: Option<String>,
    #[serde(rename = "repoName")]
    pub repo_name: Option<String>,
    #[serde(rename = "branch")]
    pub branch: Option<String>,
    #[serde(rename = "filePath")]
    pub file_path: Option<String>,
    #[serde(rename = "commitMsg")]
    pub commit_msg: Option<String>,
    #[serde(rename = "isNewBranch")]
    pub is_new_branch: Option<bool>,
    #[serde(rename = "baseBranch")]
    pub base_branch: Option<String>,
}

File: common/client/dto/infrastructure.go
package dto

// Infrastructure represents a Harness infrastructure definition
type Infrastructure struct {
	ID                string                   `json:"identifier"`
	Name              string                   `json:"name"`
	Description       string                   `json:"description,omitempty"`
	OrgIdentifier     string                   `json:"orgIdentifier"`
	ProjectIdentifier string                   `json:"projectIdentifier"`
	EnvironmentRef    string                   `json:"environmentRef"`
	Type              string                   `json:"type,omitempty"`
	DeploymentType    string                   `json:"deploymentType,omitempty"` // e.g., Kubernetes, ECS, etc.
	YAML              string                   `json:"yaml,omitempty"`
	Tags              map[string]string        `json:"tags,omitempty"`
	Variables         []map[string]interface{} `json:"variables,omitempty"`
	GitOpsEnabled     bool                     `json:"gitOpsEnabled,omitempty"`
	AccountID         string                   `json:"accountId,omitempty"`
	StoreType         string                   `json:"storeType,omitempty"`
}

// InfrastructureItem represents an item in the response list
type InfrastructureItem struct {
	Infrastructure        Infrastructure `json:"infrastructure"`
	CreatedAt             int64          `json:"createdAt,omitempty"`
	LastModifiedAt        int64          `json:"lastModifiedAt,omitempty"`
	EntityValidityDetails interface{}    `json:"entityValidityDetails"`
	GovernanceMetadata    interface{}    `json:"governanceMetadata"`
}

// InfrastructureListResponse represents the response from the list infrastructures API
type InfrastructureListResponse struct {
	Status        string      `json:"status,omitempty"`
	MetaData      interface{} `json:"metaData"`
	CorrelationID string      `json:"correlationId,omitempty"`
	Data          struct {
		Content       []InfrastructureItem `json:"content"`
		TotalPages    int                  `json:"totalPages"`
		TotalItems    int                  `json:"totalItems"`
		PageItemCount int                  `json:"pageItemCount"`
		PageSize      int                  `json:"pageSize"`
		PageIndex     int                  `json:"pageIndex"`
		Empty         bool                 `json:"empty"`
		PageToken     interface{}          `json:"pageToken"`
	} `json:"data"`
}

// InfrastructureOptions represents the options for listing infrastructures
type InfrastructureOptions struct {
	Page                  int    `json:"page,omitempty"`
	Limit                 int    `json:"limit,omitempty"`
	Sort                  string `json:"sort,omitempty"`
	Order                 string `json:"order,omitempty"`
	DeploymentType        string `json:"deploymentType,omitempty"`        // Filter by deployment type
	EnvironmentIdentifier string `json:"environmentIdentifier,omitempty"` // Filter by environment
}

// MoveInfraConfigsRequest represents the request to move infrastructure configurations
type MoveInfraConfigsRequest struct {
	InfraIdentifier       string         `json:"-"` // Required - from path parameter
	EnvironmentIdentifier string         `json:"-"` // Required
	AccountIdentifier     string         `json:"-"` // Required
	OrgIdentifier         string         `json:"-"`
	ProjectIdentifier     string         `json:"-"`
	ConnectorRef          string         `json:"-"`
	RepoName              string         `json:"-"`
	Branch                string         `json:"-"`
	FilePath              string         `json:"-"`
	CommitMsg             string         `json:"-"`
	IsNewBranch           *bool          `json:"-"`
	BaseBranch            string         `json:"-"`
	IsHarnessCodeRepo     *bool          `json:"-"`
	MoveConfigType        MoveConfigType `json:"-"` // Required - enum: "INLINE_TO_REMOTE" "REMOTE_TO_INLINE"
}

// MoveInfraConfigsResponse represents the response from the move infrastructure configs API
type MoveInfraConfigsResponse struct {
	Status        string `json:"status"`
	CorrelationId string `json:"correlationId"`
	MetaData      any    `json:"metaData"`

	Data struct {
		Identifier string `json:"identifier"`
		Success    bool   `json:"success"`
	} `json:"data,omitempty"`

	Code             string                `json:"code,omitempty"`
	Message          string                `json:"message,omitempty"`
	DetailedMessage  string                `json:"detailedMessage,omitempty"`
	ResponseMessages []HarnessErrorMessage `json:"responseMessages,omitempty"`
}
